docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+##### BYO Resources
+
+BYO resources likely will not exist as ASO resources, so determining whether or not a given expected Azure resource exists becomes less straightforward using ASO. Initially, both BYO and CAPZ-managed resources will be created in ASO with ASO's `serviceoperator.azure.com/reconcile-policy: skip` label which tells ASO not to make any changes to the corresponding Azure resource, essentially assuming the resource is BYO. Then, if a future CAPZ reconciliation reveals the Azure resource does not exist, CAPZ will change it's assumption and make ASO "adopt" the resource by setting `serviceoperator.azure.com/reconcile-policy: manage` to allow changes made to the ASO resource to be reflected in Azure.
+
+#### Installation
+
+ASO is a self-contained component published independently of CAPZ. When CAPZ is relying on ASO for critical functionality, CAPZ's installation process should also automatically install ASO.
+
+When installing CAPZ with `clusterctl init`, the Kubernetes resources defined by CAPZ's `infrastructure-components.yaml` published on the corresponding release are installed. ASO can also be [installed with a YAML manifest](https://azure.github.io/azure-service-operator/introduction/installing-from-yaml/) published for each of its releases describing its components that can be incorporated with CAPZ's `infrastructure-components.yaml`. This method installs ASO in the `azureserviceoperator-system` namespace and expects a Secret named `aso-controller-settings` to exist with Service Principal credentials. The easiest and most flexible way to customize that manifest further would likely be to generate a manifest from ASO's Helm chart with any necessary values overridden.
+
+Other than integrating with CAPZ's `infrastructure-components.yaml`, the `capz-controller-manager` container itself could assume responsibility for installing ASO by using the Helm Go library to install the ASO chart when it starts. Installing ASO this way allows CAPZ to hook in additional logic, such as to check if ASO is already installed in the cluster or to verify ASO is healthy after it's installed, which would not be possible if a static manifest was embedded in `infrastructure-components.yaml`.
+
+In terms of specific options, ASO offers both single- and multi-operator configurations. The single-operator option is the default, and assumes a single ASO installation managing resources for the whole cluster. The multi-operator method allows several ASO installations to coexist, each monitoring resources in a unique set of namespaces to enable multitenancy. ASO's single-operator configuration also allows multitenancy with more tightly-scoped credentials. Overall, a single ASO installation will likely be the easier path forward since it's consistent with CAPZ's multitenancy model and doesn't involve managing ASO's CRDs and webhooks across multiple installations.
+
+#### Caveats
+
```

Comment:
is this something that could be added to ASO in the future?

Reply:
For resource health in particular, I could see where that may never be added to ASO since it's not really a separate resource. However, it seems possible to have an ASO resource that looks something like
```yaml
kind: ResourceHealth
spec:
  id: <resource ID>
status:
  availabilityState: Available
  ...
```
where the `spec` refers to a resource whose health is reported in `status`.

Or, resource health might be able to become a part of an ASO ManagedCluster's status.

@matthchr Do you think this is worth opening an issue for in ASO?

Reply:
> @matthchr Do you think this is worth opening an issue for in ASO?

Yes.

We'll need to design what this looks like in ASO, and my gut feeling is that like the SDK, resource health won't be integrated onto resource status _by default_, but that either you can create another resource (like you posited @nojnhuh), or you can set a flag at the ASO config level or an annotation on the resource that instructs us to merge the resource health data into the status.